crates/sui-core/src/transaction_manager.rs

```diff
@@ -93,7 +95,134 @@ impl LockQueue {
+        }
+    }
+}
+
+impl CacheInner {
+    fn insert(&mut self, object: &InputKey) {
+        if let Some(version) = object.1 {
+            if let Some((previous_id, previous_version)) =
+                self.versioned_cache.push(object.0, version)
+            {
+                if previous_id == object.0
+                    && previous_version > object.1.unwrap_or(SequenceNumber::MIN)
+                {
+                    // do not allow highest known version to decrease
+                    self.versioned_cache.put(object.0, previous_version);
+                } else {
+                    self.metrics
+                        .transaction_manager_object_cache_evictions
+                        .inc();
+                }
+            }
+        } else if let Some((previous_id, _)) = self.unversioned_cache.push(object.0, ()) {
+            if previous_id != object.0 {
+                self.metrics
+                    .transaction_manager_package_cache_evictions
+                    .inc();
+            }
+        }
+    }
+
+    // Returns Some(true/false) for a definitive result. Returns None if the caller must defer to
+    // the db.
+    fn is_object_available(&mut self, object: &InputKey) -> Option<bool> {
+        if let Some(version) = object.1 {
+            if let Some(current) = self.versioned_cache.get(&object.0) {
+                self.metrics.transaction_manager_object_cache_hits.inc();
+                Some(*current >= version)
+            } else {
+                self.metrics.transaction_manager_object_cache_misses.inc();
+                None
+            }
+        } else {
+            self.unversioned_cache
+                .get(&object.0)
+                .tap_some(|_| self.metrics.transaction_manager_package_cache_hits.inc())
+                .tap_none(|| self.metrics.transaction_manager_package_cache_misses.inc())
+                .map(|_| true)
+        }
+    }
+}
+
+struct AvailableObjectsCache {
+    bounded_cache: CacheInner,
+    unbounded_cache: Option<CacheInner>,
+
+    unbounded_cache_enabled: usize,
+    metrics: Arc<AuthorityMetrics>,
+}
+
+impl AvailableObjectsCache {
+    fn new(metrics: Arc<AuthorityMetrics>) -> Self {
+        Self {
+            bounded_cache: CacheInner::bounded(100000, metrics.clone()),
+            unbounded_cache: None,
+            unbounded_cache_enabled: 0,
+            metrics,
+        }
+    }
+
+    fn enable_unbounded_cache(&mut self) {
+        if self.unbounded_cache_enabled == 0 {
+            assert!(self.unbounded_cache.is_none());
+            self.unbounded_cache = Some(CacheInner::unbounded(self.metrics.clone()));
+        }
+        self.unbounded_cache_enabled += 1;
+    }
+
+    fn disable_unbounded_cache(&mut self) {
+        assert!(self.unbounded_cache_enabled > 0);
+        assert!(self.unbounded_cache.is_some());
+        self.unbounded_cache_enabled -= 1;
+        if self.unbounded_cache_enabled == 0 {
+            self.unbounded_cache = None;
```

Comment:
If there will be both bounded and unbounded caches, should we also dump the content from unbounded cache back into the bounded cache? Otherwise it seems we are evicting hot entries.

Reply:
i think i was inserting into the bounded cache no matter what, but there is now only one cache so it doesn't matter